datafusion/core/src/datasource/listing/table.rs

```diff
@@ -736,13 +736,16 @@ impl TableProvider for ListingTable {
 
     async fn scan(
         &self,
-        state: &SessionState,
+        state: &dyn CatalogSession,
         projection: Option<&Vec<usize>>,
         filters: &[Expr],
         limit: Option<usize>,
     ) -> Result<Arc<dyn ExecutionPlan>> {
-        let (mut partitioned_file_lists, statistics) =
-            self.list_files_for_scan(state, filters, limit).await?;
+        // TODO remove downcast_ref from here?
```

Comment:
Yes, definitely we should remove the downcast (likely by adding some more functions to the `CatalogSession` trait) . But I also think we could do that as a follow on PR with minimal impact

Also this pattern shows that existing code could always use this pattern to downcast `&dy CatalogSession` to `&SessionState` which gives us an easy out for all migrations

Reply:
this is a good point about the migrations

Reply:
filed  https://github.com/apache/datafusion/issues/11600 for downcasts